src/lib/ucan.ts

```diff
@@ -11,70 +9,63 @@ export type Validation = {
-  const notValidYet = isNotValidYet(token)
-  const active = !ucan.isExpired(token)
-  const valid = await validateSignature(token)
-  const [validIssuer, validProofs] = await validateProofs(token.payload.prf, token.payload.iss)
+export const validate = async (token: string): Promise<{ validation: Validation; ucan: Ucan } | null> => {
+  const parsed = await decode(token)
+  const [header, payload, signature] = token.split(".")
+  if (parsed == null || header == null || payload == null || signature == null) {
+    return null
+  }
+  const notValidYet = ucan.isTooEarly(parsed)
+  const active = !ucan.isExpired(parsed)
+  const valid = await ucan.verifySignatureUtf8(`${header}.${payload}`, signature, parsed.payload.iss)
+  const { validIssuer, validProofs } = await validateProofs(parsed.payload.prf, parsed.payload.iss)
 
   return {
-    notValidYet,
-    active,
-    valid,
-    validIssuer,
-    validProofs
+    validation: {
+      notValidYet,
+      active,
+      valid,
+      validIssuer,
+      validProofs,
+    },
+    ucan: parsed,
   }
 }
 
-const validateSignature = async (token: Ucan): Promise<boolean> => {
-  const encodedHeader = ucan.encodeHeader(token.header)
-  const encodedPayload = ucan.encodePayload(token.payload)
-
-  const data = uint8arrays.fromString(`${encodedHeader}.${encodedPayload}`)
-  const signature = uint8arrays.fromString(token.signature, 'base64urlpad')
-
-  return ucan.verifySignature(data, signature, token.payload.iss)
-}
-
-const validateProofs = (proofs: string[], delegate: string): Promise<[boolean, boolean]> => {
-  const proofValidations = proofs.map(async proof => await validateProof(proof, delegate))
-
-  return Promise.all(proofValidations).then(promisedValidations =>
-    promisedValidations.reduce(([validIssuer, validProofs], validation) =>
-      [validIssuer && validation.validIssuer, validProofs && validation.validProof],
-      [true, true]
-    )
+const validateProofs = async (proofs: string[], delegate: string): Promise<{ validIssuer: boolean; validProofs: boolean }> => {
+  const promisedValidations = await Promise.all(proofs.map(proof => validateProof(proof, delegate)))
+  return promisedValidations.reduce(
+    ({ validIssuer, validProofs }, validation) => ({
+      validIssuer: validIssuer && validation.validIssuer,
+      validProofs: validProofs && validation.validProof,
+    }),
+    { validIssuer: true, validProofs: true }
   )
 }
 
 const validateProof = async (proof: string, delegate: string): Promise<{ validIssuer: boolean; validProof: boolean }> => {
-  const token = decode(proof)
+  const token = await decode(proof)
 
-  const validIssuer: boolean = token?.payload.aud === delegate ? true : false
-  let validProof: boolean
+  let validProof: boolean = false
 
   if (token !== null) {
     try {
-      validProof = await ucan.isValid(token)
+      await ucan.validate(proof)
+      validProof = true
     } catch {
-      validProof = false
```

Comment:
Eslint doesn't like that we have an empty catch block here. We could set `validProof` to `false` here or relax that eslint rule.

I'd be alright with either. What do you think?

Reply:
Actually reads much better without that removed. Not sure why I did that :smile: